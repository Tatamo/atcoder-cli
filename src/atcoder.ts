import {SessionInterface} from "./session";
import querystring from "query-string";
import {Contest, Task} from "./project";

const ATCODER_BASE_URL = "https://atcoder.jp/";

export class AtCoder {
	static get base_url(): string {
		return ATCODER_BASE_URL;
	}

	static get login_url(): string {
		return `${AtCoder.base_url}login`;
	}

	static getContestURL(contest: string) {
		return `${AtCoder.base_url}contests/${contest}`;
	}

	static getTaskURL(contest: string, task?: string) {
		return `${AtCoder.getContestURL(contest)}/tasks${task === undefined ? "" : `/${task}`}`;
	}

	private session: SessionInterface;
	// null:未検査 true/false: ログインしているかどうか
	private _login: boolean | null;

	constructor(session: SessionInterface) {
		this.session = session;
		this._login = null;
	}

	/**
	 * ログインしているか調べる
	 * @param force default=false trueならキャッシュを使わずちゃんと調べる
	 */
	async checkSession(force: boolean = false): Promise<boolean> {
		// 以前取得済みならいちいち接続して確かめない
		if (this._login !== null && !force) return this._login;
		return this._login = await this.check();
	}

	/**
	 * アクセスしてログインしている状態かどうかを取得する(結果をキャッシュしない)
	 */
	private async check(): Promise<boolean> {
		// ログインせず提出ページにアクセスするとコンテストトップに飛ばされることを利用する
		const url = `${AtCoder.getContestURL("abc001")}/submit`;
		// リダイレクトを無効化・302コードを容認して通信
		const response = await this.session.get(url, {
			maxRedirects: 0,
			validateStatus: (status) => (status >= 200 && status < 300) || status === 302
		});
		// リダイレクトされなければログインしている
		return response.status !== 302;
	}

	/**
	 * ログイン処理します
	 * あまりパスワード文字列を引き回したくないので、この中で標準入力からユーザー名とパスワードを尋ねる
	 */
	async login(options: {username?: string, password?: string}): Promise<boolean> {
		return this.session.transaction(async ()=> {
			const {csrf_token} = await this.getCSRFToken();

			// ユーザーネームとパスワードを入力させる
			const inquirer = await import("inquirer");
			const username : string = options.username || (await inquirer.prompt([{
				type: "input",
				message: "username:",
				name: "username"
			}]) as {username: string})["username"];
			const password : string = options.password || (await inquirer.prompt([{
				type: "password",
				message: "password:",
				name: "password"
			}]) as {password: string})["password"];

			const response = await this.session.post(
				AtCoder.login_url,
				querystring.stringify({username, password, csrf_token}),
				{
					maxRedirects: 0,
					validateStatus: (status) => (status >= 200 && status < 300) || status === 302,
				}
			)


			// ログインページ以外にリダイレクトされていればログイン成功とみなす
			const result = response.headers.location !== "/login";
			if (result) {
				// ログインに成功していた場合はセッション情報を保存する
				await response.saveSession()
			}
			return result;
		});
	}

	/**
	 * ログインページにアクセスしてCSRFトークンを取得
	 */
	private async getCSRFToken(): Promise<{ csrf_token: string }> {
		const {JSDOM} = await import("jsdom");
		// cookieなしでログインページにアクセス
		const response = await this.session.get(AtCoder.login_url, {headers: {Cookie: ""}});

		const {document} = new JSDOM(response.data).window;
		const input: HTMLInputElement = (document.getElementsByName("csrf_token")[0]) as HTMLInputElement;

		await response.saveSession();
		return {csrf_token: input.value};
	}

	async logout(): Promise<void> {
		await this.session.removeSession();
		this._login = null;
	}

	/**
	 * コンテストIDからコンテストの情報を取得
	 * @param id
	 * @throws Error
	 */
	async contest(id: string): Promise<Contest> {
		const url = AtCoder.getContestURL(id);
		// コンテストが見つからない場合エラーとなるがハンドルせず外に投げる
		const response = await this.session.get(url);
		const {JSDOM} = await import("jsdom");
		const {document} = new JSDOM(response.data).window;
		const regexp = /^(.*) - AtCoder$/;
		const title = regexp.test(document.title) ? regexp.exec(document.title)![1] : document.title;
		return {id, title, url};
	}

	/**
	 * 問題一覧を取得
	 * @param contest_id
	 * @throws Error
	 */
	async tasks(contest_id: string): Promise<Array<Task>> {
		// コンテストが見つからない場合エラーとなるがハンドルせず外に投げる
		const response = await this.session.get(AtCoder.getTaskURL(contest_id));

		const {JSDOM} = await import("jsdom");
		const {document} = new JSDOM(response.data).window;
		// very very ad-hoc and not type-safe section
		const tbody = document.querySelector("#main-div .row table>tbody");
		if (tbody === null) return [];
		const tasks: Array<Task> = [];
		for (const tr of tbody.querySelectorAll("tr")) {
			// tr>td>a
			const id: string = tr.children[0].children[0].getAttribute("href")!.split("/").pop()!;
			const label: string = (tr.children[0].children[0] as HTMLAnchorElement).text;
			const title: string = (tr.children[1].children[0] as HTMLAnchorElement).text;
			const url: string = `${AtCoder.base_url}${tr.children[0].children[0].getAttribute("href")!.substring(1)}`;
			tasks.push({id, label, title, url});
		}
		return tasks;
	}

	/**
	 * 単一の問題を取得
	 * @param contest_id
	 * @param task_id
	 * @throws Error
	 */
	async task(contest_id: string, task_id: string): Promise<Task> {
		const tasks = await this.tasks(contest_id);
		for (const task of tasks) {
			if (task.id === task_id) return task;
		}
		throw new Error(`Task ${task_id} not found.`);
	}
}
