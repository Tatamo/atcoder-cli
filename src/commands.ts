import {AtCoder} from "./atcoder";
import {OnlineJudge} from "./facade/oj";
import {Contest, Task, detectTaskByPath, findProjectJSON, formatTaskDirname, saveProjectJSON, init, installTask, formatContestDirname} from "./project";
import getConfig, {defaults, getConfigDirectory} from "./config";
import {getTemplate, getTemplates, Template} from "./template";
import {getAtCoder} from "./di";

export async function login(options: {username?: string, password?: string}) {
	const atcoder = await getAtCoder();
	console.log(await atcoder.login(options) ? "OK" : "login failed");
}

export async function logout() {
	const atcoder = await getAtCoder();
	await atcoder.logout();
	console.log("login session aborted.");
}

export async function session() {
	const atcoder = await getAtCoder();
	console.log("check login status...");
	console.log(await atcoder.checkSession() ? "OK" : "not login");
}

export async function contest(contest_id: string | undefined, options: {title?: boolean, id?: boolean, url?: boolean}) {
	const f_id = options.id === true;
	const f_title = options.title === true;
	const f_url = options.url === true;
	const format = ({ id, title, url }: Contest) => formatAsShellOutput([((f_id || f_title || f_url) ? [f_id ? SGR(id, 37) : null, f_title ? SGR(title, 32, 1) : null, f_url ? url : null].filter(e => e !== null) : [SGR(id, 37), SGR(title, 32, 1), url]) as Array<string>]);
	if (contest_id === undefined) {
		// idが与えられていない場合、プロジェクトファイルを探してコンテスト情報を表示
		try {
			const {data: {contest}} = await findProjectJSON();
			console.log(format(contest));
		} catch (e) {
			console.error(e.message);
		}
	}
	else {
		try {
			const atcoder = await getAtCoder();
			if (!await atcoder.checkSession()) await atcoder.login({});
			const contest = await atcoder.contest(contest_id);
			console.log(format(contest));
		} catch {
			console.error(`contest "${contest_id}" not found.`);
		}
	}
}

export async function task(contest_id: string | undefined, task_id: string | undefined, options: {label?: boolean, title?: boolean, id?: boolean, url?: boolean}) {
	const f_id = options.id === true;
	const f_title = options.title === true;
	const f_url = options.url === true;
	const f_label = options.label === true;

	// in case all flags are false (no print options specified), the whole (id, label, title, url string) is printed
	const format = ({id, label, title, url}: Task) => formatAsShellOutput([((f_id || f_label || f_title || f_url) ? [f_id ? SGR(id, 37) : null, f_label ? SGR(label, 32) : null, f_title ? SGR(title, 32, 1) : null, f_url ? url : null].filter(e => e !== null) : [SGR(id, 37), SGR(label, 32), SGR(title, 32, 1), url]) as Array<string>]);
	if (contest_id === undefined && task_id === undefined) {
		// idが与えられていない場合、プロジェクトファイルを探す
		try {
			const {task} = await detectTaskByPath();
			if (task === null) {
				console.error("failed to find the task.");
				return;
			}
			console.log(format(task));
		} catch (e) {
			console.error(e.message);
		}
	}
	else if (contest_id !== undefined && task_id !== undefined) {
		try {
			const atcoder = await getAtCoder();
			if (!await atcoder.checkSession()) await atcoder.login({});
			const task = await atcoder.task(contest_id, task_id);
			console.log(format(task));
		} catch {
			console.error(`task "${task_id}" of contest "${contest_id}" not found.`);
		}
	}
	else {
		console.error("error: specify both the contest id and the task id.")
	}
}

export async function tasks(contest_id: string | undefined, options: {id?: boolean}) {
	const f_id = options.id === true;
	const format = (tasks: Array<Task>) => formatAsShellOutput(tasks.map(({id, label, title, url}) => [f_id ? SGR(id, 37) : null, SGR(label, 32), SGR(title, 32, 1), url].filter(e => e !== null) as Array<string>));
	if (contest_id === undefined) {
		// idが与えられていない場合、プロジェクトファイルを探す
		try {
			const {data: {tasks}} = await findProjectJSON();
			console.log(format(tasks));
		} catch (e) {
			console.error(e.message);
		}
	}
	else {
		try {
			const atcoder = await getAtCoder();
			if (!await atcoder.checkSession()) await atcoder.login({});
			const tasks = await atcoder.tasks(contest_id);
			console.log(format(tasks));
		} catch {
			console.error(`contest "${contest_id}" not found.`);
		}
	}
}

export async function url(contest_id: string | undefined, task_id: string | undefined, options: {check?: boolean}) {
	const f_check = options.check === true;
	if (contest_id !== undefined && task_id !== undefined) {
		if (f_check) {
			const atcoder = await getAtCoder();
			try {
				const tasks = await atcoder.tasks(contest_id);
				// Task一覧から一致するURLを探す
				for (const {url} of tasks) {
					if (url === AtCoder.getTaskURL(contest_id, task_id)) {
						console.log(url);
						return;
					}
				}
				// なかった
				console.error(`task "${task_id}" not found.`);
			} catch {
				console.error(`contest "${contest_id}" not found.`);
			}
		}
		else {
			// URLの妥当性をチェックしない
			console.log(AtCoder.getTaskURL(contest_id, task_id));
		}
	}
	else if (contest_id !== undefined && task_id === undefined) {
		if (f_check) {
			const atcoder = await getAtCoder();
			try {
				// コンテストページの存在確認
				const {url} = await atcoder.contest(contest_id);
				console.log(url);
			} catch {
				console.error(`contest "${contest_id}" not found.`);
			}
		}
		else {
			// URLの妥当性をチェックしない
			console.log(AtCoder.getContestURL(contest_id));
		}
	}
	else {
		console.log(AtCoder.base_url);
	}
}

export async function format(format_string: string, contest_id: string, task_id?: string) {
	const atcoder = await getAtCoder();
	if (!await atcoder.checkSession()) await atcoder.login({});
	if (task_id === undefined) {
		// コンテスト情報のみを使用
		try {
			const contest = await atcoder.contest(contest_id);
			console.log(formatContestDirname(format_string, contest));
		} catch (e) {
			console.error(e.toString());
		}
	}
	else {
		// コンテスト・問題情報を使用
		try {
			const [contest, tasks] = await Promise.all([atcoder.contest(contest_id), atcoder.tasks(contest_id)]);
			// 問題番号を調べる
			let index = -1;
			for (let i = 0; i < tasks.length; i++) {
				if (tasks[i].id === task_id) {
					index = i;
					break;
				}
			}
			if (index < 0) {
				console.error(`task ${task_id} not found.`);
				return;
			}
			try {
				console.log(formatTaskDirname(format_string, tasks[index], index, contest));
			} catch (e) {
				console.error(e.toString());
			}
		} catch {
			// TODO: もう少し良いエラーハンドリングができないものか
			console.error("failed to get contest information.");
		}
	}
}

export async function submit(filename: string | undefined, facade_options: Array<string>, options: {task?: string, contest?: string, skipFilename?: boolean}) {
	let contest_id = options.contest;
	let task_id = options.task;
	const f_skip_filename = options.skipFilename === true;

	// treat filename as a first facade option and assume that filename is not given
	if (f_skip_filename && filename !== undefined) {
		facade_options.unshift(filename);
		filename = undefined;
	}
	if (filename === undefined || contest_id === undefined || task_id === undefined) {
		// ファイル名、タスク、コンテストのいずれかが未指定の場合、カレントディレクトリのパスから提出先を調べる
		const {contest, task} = await detectTaskByPath();
		if (filename === undefined && task !== null && task.directory !== undefined) filename = task.directory.submit;
		if (contest_id === undefined && contest !== null) contest_id = contest.id;
		if (task_id === undefined && task !== null) task_id = task.id;

		// 結局特定できなかった
		if (filename === undefined) {
			console.error(`the program file to submit is not found.`);
			return;
		}
		if (contest_id === undefined || task_id === undefined) {
			console.error(`cannot find the ${task_id === undefined ? "task" : "contest"} to submit.`);
			console.error(`add ${task_id === undefined ? "-t" : "-c"} flag to specify it.`);
			return;
		}
	}

	if (!await OnlineJudge.checkAvailable()) {
		console.error("online-judge-tools is not available.");
		return;
	}

	// URLの妥当性をチェック
	const atcoder = await getAtCoder();
	const url: string | null = (await atcoder.task(contest_id, task_id).catch(() => ({url: null}))).url;
	if (url === null) {
		console.error(`Task ${AtCoder.getTaskURL(contest_id, task_id)} not found.`);
		return;
	}
	console.log(`submit to: ${url}`);
	// 提出
	await OnlineJudge.call(["s", url, filename, ...facade_options]);
}

export async function checkOJAvailable() {
	const available = await OnlineJudge.checkAvailable();
	const path = await OnlineJudge.getPath();
	console.log(`online-judge-tools is ${available ? "" : "not "}available. ${available ? "found at:" : ""}`);
	if (available) {
		console.log(path);
	}
}

export async function configDir() {
	console.log(await getConfigDirectory());
}

export async function config(key: string | undefined, value: string | undefined, options: {D?: boolean}) {
	if (options.D) {
		await deleteGlobalConfig(key);
	}
	else if (key !== undefined && value !== undefined) {
		await setGlobalConfig(key, value);
	}
	else {
		await getGlobalConfig(key);
	}
}

async function getGlobalConfig(key?: string) {
	const conf = await getConfig();
	if (key === undefined) {
		for (const key of Object.keys(defaults)) {
			console.log(undef2empty`${key}: ${conf.get(key)}`)
		}
		return;
	}
	if (!(key in defaults)) {
		console.error(`invalid option "${key}".`);
		return;
	}
	console.log(undef2empty`${conf.get(key)}`);
}

async function setGlobalConfig(key: string, value: string) {
	const conf = await getConfig();
	if (!(key in defaults)) {
		console.error(`invalid option "${key}".`);
		return;
	}
	conf.set(key, value);
	console.log(undef2empty`${key} = ${conf.get(key)}`);
}

async function deleteGlobalConfig(key?: string) {
	const conf = await getConfig();
	if (key !== undefined) {
		if (!(key in defaults)) {
			console.error(`invalid option "${key}".`);
			return;
		}
		conf.delete(key);
		console.log(`option "${key}" is set back to default.`);
	}
	else {
		console.error("option key is not specified.");
	}
}

/**
 * テンプレート文字列に挿入された式がundefinedであった場合に"undefined"のかわりに空文字列に変換する
 * @param strings
 * @param values
 */
function undef2empty(strings: TemplateStringsArray, ...values: Array<any>): string {
	values = values.map(value => value !== undefined ? value : "");
	return String.raw(strings, ...values);
}

type Choices = "inquire" | "all" | "none" | "rest" | "next"

/**
 * --choiceオプションに渡される値として適切かどうかを判定する
 * @param choice
 */
function checkValidChoiceOption(choice: any): choice is Choices {
	switch (choice) {
		case "inquire":
		case "all":
		case "none":
		case "rest":
		case "next":
			return true;
	}
	return false;
}

async function getTemplateFromOption(template?: string | boolean): Promise<Template | undefined> {
	// --no-templateオプションが指定された場合は何も選ばない
	if (template === false) return undefined;
	// 未指定の場合はコンフィグよりデフォルト値を取得
	if (template === undefined || template === true) template = (await getConfig()).get("default-template");
	// デフォルト値も指定されていなければ何も選ばない
	if (template === "") return undefined;
	return await getTemplate(template).catch((e) => {
		throw new Error(`Failed to load template "${template}".\n  ${e}`);
	});
}

export async function setup(contest_id: string, options: {choice: Choices, force?: boolean, contestDirnameFormat?: string, taskDirnameFormat?: string, template?: string | boolean, tests?: boolean}) {
	try {
		const template = await getTemplateFromOption(options.template);
		const {contest} = await init(contest_id, template, options);
		console.log(`create project of ${contest.title}`);
		await add(options);
	} catch (e) {
		console.error(e.message);
	}
}

export async function add(options: {choice?: Choices, force?: boolean, taskDirnameFormat?: string, template?: string | boolean, tests?: boolean}) {
	try {
		const {path, data} = await findProjectJSON();
		const {contest, tasks} = data;
		const choices = await (async (c: string | undefined) => {
			// choiceオプションが設定されていない場合はコンフィグからデフォルト値を取得
			let flg_default_choice = false;
			if (c === undefined) {
				c = (await getConfig()).get("default-task-choice");
				flg_default_choice = true;
			}
			// 有効な値が与えられているかどうかバリデーションを行う
			if (!checkValidChoiceOption(c)) {
				throw new Error(`Invalid option is given with --choice. (inquire/all/none/rest/next)${flg_default_choice ? "\nIf you did not set the --choice option, check the \"default-task-choice\" option in global config by using `acc config`." : ""}`);
			}
			return await selectTasks(tasks, c, options.force);
		})(options.choice);
		const template = await getTemplateFromOption(options.template);
		// 更新があった問題の数を数えておく
		let count = 0;
		for (const {index, task} of choices) {
			// forceオプションが設定されていない場合、既にディレクトリが存在する問題はスキップする
			if (options.force !== true && task.directory !== undefined) continue;
			const format = options.taskDirnameFormat !== undefined ? options.taskDirnameFormat : (await getConfig()).get("default-task-dirname-format");
			const dirname = formatTaskDirname(format, task, index, contest);
			// 新しいTaskが返ってくるので、もともとの配列の要素を更新する
			tasks[index] = await installTask({task, index, contest, template}, dirname, path, {tests: options.tests});
			count++;
		}
		if (count === 0) {
			// 問題が何も更新されなかった場合、その旨を通知する
			console.error(SGR("Skip: no task directories were created. use --force option to overwrite existent directories.", 33));
		}
		// 更新されたContestProjectをファイルに書き出し
		await saveProjectJSON(Object.assign(data, {tasks}));
	} catch (e) {
		console.error(SGR(e.message, 33));
	}
}

export async function selectTasks(tasks: Array<Task>, choice: Choices, force: boolean = false): Promise<Array<{index: number, task: Task}>> {
	switch (choice) {
		case "inquire":
			return await inquireTasks(tasks, force);
		case "all":
			return tasks.map((task, index) => ({index, task}));
		case "none":
			return [];
		case "rest":
			return tasks.filter(task => task.directory === undefined).map((task, index) => ({index, task}));
		case "next":
			const next = getNextTask2Install(tasks);
			return next !== null ? [next] : [];
		default:
			throw new Error(`mode "${choice}" is not defined"`);
	}
}

export async function inquireTasks(tasks: Array<Task>, force: boolean = false): Promise<Array<{index: number, task: Task}>> {
	const inquirer = await import("inquirer");
	// まだディレクトリが作成されていない問題を一つだけ選択状態にしておく
	const next = getNextTask2Install(tasks);
	if (!force && next === null) throw new Error("all tasks are already installed. use --force option to overwrite them.");
	return (await inquirer.prompt([{
		type: "checkbox",
		message: "select tasks",
		name: "tasks",
		choices: tasks.map((task, index) => ({
			name: `${task.label} ${task.title}`,
			value: {index, task},
			disabled: !force && task.directory !== undefined ? () => "already installed" : () => false,
			checked: next !== null && index === next.index
		}))
	}]) as {tasks: Array<{index: number, task: Task}>}).tasks;
}

/**
 * まだディレクトリが作成されていない問題のうち、最も上のものを得る
 * すべての問題のディレクトリが作成済みの場合はnullを返す
 * @param tasks
 */
function getNextTask2Install(tasks: Array<Task>): {index: number, task: Task} | null {
	for (let i = 0; i < tasks.length; i++) {
		if (tasks[i].directory === undefined) {
			return {index: i, task: tasks[i]};
		}
	}
	return null;
}

export async function getTemplateList() {
	console.error(SGR(`search template directories in ${await getConfigDirectory()}`, 37));
	const templates = await getTemplates();
	console.log(formatAsShellOutput([[SGR("NAME", 1), "SUBMIT-PROGRAM"]].concat(templates.map(template => [SGR(template.name, 1), template.task.submit]))));
}

/**
 * 文字列の2次元配列を受け取り、行ごとにいい感じのスペースで結合して返す
 * @param data
 */
function formatAsShellOutput(data: Array<Array<string>>): string {
	const padding = "  ";
	const max_lengths = data.map(arr => arr.map(str => str.length)).reduce((a, b) => a.map((v, i) => Math.max(v, b[i])));
	return data.map(line => line.reduceRight((p, c, i) => c.padEnd(max_lengths[i]) + padding + p)).join("\n");
}

/**
 * プロセスがTTYで実行されているならSGRでスタイルを適用する
 * @param str 適用する文字列
 * @param codes コード番号(可変長)
 */
function SGR(str: string, ...codes: Array<number>) {
	if (!process.stdout.isTTY) return str;
	return `${codes.map(code => `\x1b[${code}m`).join("")}${str}\x1b[0m`;
}
