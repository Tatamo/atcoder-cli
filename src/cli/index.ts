import commander from "commander";
import {name, version} from "../../package.json";
import * as commands from "../commands";
import * as help from "../help";
import updateNotifier from "update-notifier";

updateNotifier({pkg: {name, version}}).notify({isGlobal: true});

commander
	.version(version, "-v, --version");

commander
	.command("new <contest-id>")
	.alias("n")
	.action(commands.setup)
	.option('-c, --choice <choice>', "how to choice tasks to add", /^(inquire|all|none|rest|next)$/i)
	.option("-f, --force", "ignore existent directories")
	.option("-d, --contest-dirname-format <format>", "specify the format to name contest directory. defaults to \"{ContestID}\"")
	.option("-t, --task-dirname-format <format>", "specify the format to name task directories. defaults to \"{tasklabel}\"")
	.option("--no-tests", "skip downloading sample cases by using online-judge-tools")
	.option("--template <name>", "specify the provisioning template")
	.option("--no-template", "do not use templates, even if specified by global config")
	.description("create new contest project directory")
	.on("--help", () => {
		console.log("");
		console.log(help.task_choices);
	});

commander
	.command("add")
	.alias("a")
	.action(commands.add)
	.option('-c, --choice <choice>', "how to choice tasks to add", /^(inquire|all|none|rest|next)$/i)
	.option("-f, --force", "ignore existent directories")
	.option("-t, --task-dirname-format <format>", "specify the format to name task directories. defaults to \"{tasklabel}\"")
	.option("--no-tests", "skip downloading sample cases by using online-judge-tools")
	.option("--template <name>", "specify the provisioning template")
	.option("--no-template", "do not use templates, even if specified by global config")
	.description("add new directory for the task in the project directory")
	.on("--help", () => {
		console.log("");
		console.log(help.task_choices);
	});

commander
	.command("submit [filename] [facade-options...]")
	.alias("s")
	.option("-c, --contest <contest-id>", "specify contest id to submit")
	.option("-t, --task <task-id>", "specify task id to submit")
	.option("-s, --skip-filename", "specify that filename is not given (the first argument will be parsed as not a filename, but a facade option)")
	.action(commands.submit)
	.description("submit the program")
	.on("--help", () => {
		console.log("");
		console.log(help.submit_facade_options);
	});

commander
	.command("login")
	.option("-u, --username <username>", "your username")
	.option("-p, --password <password>", "your password")
	.action(commands.login)
	.description("login to AtCoder");

commander
	.command("logout")
	.action(commands.logout)
	.description("delete login session information");

commander
	.command("session")
	.action(commands.session)
	.description("check login or not");

commander
	.command("contest [contest-id]")
	.action(commands.contest)
	.option("-t, --title", "show contest title")
	.option("-i, --id", "show contest id")
	.option("-u, --url", "show contest url")
	.description("get contest information");

commander
	.command("task [contest-id] [task-id]")
	.action(commands.task)
	.option("-l, --label", "show task label")
	.option("-t, --title", "show task title")
	.option("-i, --id", "show task id")
	.option("-u, --url", "show task url")
	.description("get task");

commander
	.command("tasks [contest-id]")
	.action(commands.tasks)
	.option("-i, --id", "show task id")
	.description("get tasks");

commander
	.command("url [contest] [task]")
	.option("-c, --check", "check the specified contest and/or task id is valid")
	.action(commands.url)
	.description("get contest or task URL");

commander
	.command("format <format> <contest-id> [task-id]")
	.action(commands.format)
	.description("format string with contest and/or task information.")
	.on("--help", () => {
		console.log("");
		console.log(help.format_strings);
	});

commander
	.command("check-oj")
	.action(commands.checkOJAvailable)
	.description("check whether online-judge-tools related functions are available or not")
	.on("--help", () => {
		console.log("");
		console.log(help.online_judge_tools);
	});

commander
	.command("config [key] [value]")
	.option("-d", "delete the option value and set back to the default")
	.action(commands.config)
	.description("get or set values of global options")
	.on("--help", () => {
		console.log("");
		console.log(help.global_config);
	});

commander
	.command("config-dir")
	.action(commands.configDir)
	.description("get the path of atcoder-cli config directory");

commander
	.command("templates")
	.action(commands.getTemplateList)
	.description("show user templates in the config directory")
	.on("--help", () => {
		console.log("");
		console.log(help.provisioning_templates);
	});


commander.on("--help", () => {
	console.log("");
	console.log(help.default_help);
});

// error on unknown commands
commander.on("command:*", function () {
	console.error('Invalid command: %s\nUse `acc --help` for a list of available commands.', commander.args.join(' '));
});

commander.parse(process.argv);
