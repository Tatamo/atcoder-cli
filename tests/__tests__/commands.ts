import { login, logout, session, contest } from "../../src/commands";
import { AtCoder } from "../../src/atcoder";
import { getAtCoder } from "../../src/di";
import { Contest, ContestProject, findProjectJSON, Task } from "../../src/project"
jest.mock("../../src/di");
jest.mock("../../src/project")

const dummy_contest: Contest = {
    id: "aic987",
    title: "AtCoder Imaginary Contest 987",
    url: AtCoder.getContestURL("aic987")
};

const dummy_task01: Task = {
    id: "aic987_a",
    label: "A",
    title: "This is Problem",
    url: AtCoder.getTaskURL("aic987", "aic987_a")
};

const dummy_task02: Task = {
    id: "aic987_b",
    label: "B",
    title: "Next Problem",
    url: AtCoder.getTaskURL("aic987", "aic987_b")
};

const dummy_task03: Task = {
    id: "aic987_c",
    label: "C",
    title: "The Test Cases",
    url: AtCoder.getTaskURL("aic987", "aic987_c")
};

const dummy_task04: Task = {
    id: "aic987_d",
    label: "D",
    title: "Imaginary Problem",
    url: AtCoder.getTaskURL("aic987", "aic987_d")
};

const dummy_contest_project: ContestProject = {
    contest: dummy_contest,
    tasks: [dummy_task01, dummy_task02, dummy_task03, dummy_task04]
};

describe("Commands", () => {
    let atcoder: AtCoder;
    let log = "";
    let elog = "";
    beforeAll(async () => {
        jest.spyOn(console, "log").mockImplementation(s => log += `${s}\n`);
        jest.spyOn(console, "error").mockImplementation(s => elog += `${s}\n`);
        atcoder = jest.genMockFromModule<{ AtCoder: AtCoder }>("../../src/atcoder").AtCoder;
        // @ts-ignore: dynamically added method for test
        getAtCoder.mockResolvedValue(atcoder);
    });
    beforeEach(() => {
        // clear call histories of mock functions
        jest.clearAllMocks();
        log = "";
        elog = "";
    });
    describe("login()", () => {
        test("login success", async () => {
            atcoder.login = jest.fn(async () => true);

            await login();
            expect(log).toMatchSnapshot();
            expect(atcoder.login).toBeCalledTimes(1);
        })
        test("login fail", async () => {
            atcoder.login = jest.fn(async () => false);

            await login();
            expect(log).toMatchSnapshot();
            expect(atcoder.login).toBeCalledTimes(1);
        })
    });
    describe("logout()", () => {
        test("logout", async () => {
            atcoder.logout = jest.fn();

            await logout();
            expect(log).toMatchSnapshot();
            expect(atcoder.logout).toBeCalledTimes(1);
        })
    });
    describe("session()", () => {
        test("logged-in", async () => {
            atcoder.checkSession = jest.fn(async () => true);

            await session();
            expect(log).toMatchSnapshot();
            expect(elog).toMatchSnapshot();
            expect(atcoder.checkSession).toBeCalledTimes(1);
        })
        test("not logged-in", async () => {
            atcoder.checkSession = jest.fn(async () => false);

            await session();
            expect(log).toMatchSnapshot();
            expect(elog).toMatchSnapshot();
            expect(atcoder.checkSession).toBeCalledTimes(1);
        })
    });
    describe("contest()", () => {
        describe("contest id is not given", () => {
            test("find project json found, without id", async () => {
                // @ts-ignore: dynamically added method for test
                findProjectJSON.mockResolvedValueOnce({ path: "/dummy/path", data: dummy_contest_project });
                await contest(undefined, {});
                expect(log).toMatchSnapshot();
                expect(elog).toMatchSnapshot();
                expect(findProjectJSON).toBeCalledTimes(1);
            })
            test("find project json found, with id", async () => {
                // @ts-ignore: dynamically added method for test
                findProjectJSON.mockResolvedValueOnce({ path: "/dummy/path", data: dummy_contest_project });
                await contest(undefined, { id: true });
                expect(log).toMatchSnapshot();
                expect(elog).toMatchSnapshot();
                expect(findProjectJSON).toBeCalledTimes(1);
            })
            test("project json not found", async () => {
                // @ts-ignore: dynamically added method for test
                findProjectJSON.mockRejectedValueOnce(new Error("contest.acc.json not found."));
                await contest(undefined, {});
                expect(log).toMatchSnapshot();
                expect(elog).toMatchSnapshot();
                expect(findProjectJSON).toBeCalledTimes(1);
            });
        });
        describe("contest id is given", () => {
            test("contest found, without id", async () => {
                atcoder.checkSession = jest.fn(async () => true);
                atcoder.contest = jest.fn(async _ => dummy_contest);
                await contest("aic987", {});
                expect(log).toMatchSnapshot();
                expect(elog).toMatchSnapshot();
                expect(findProjectJSON).toBeCalledTimes(0);
                expect(atcoder.contest).toBeCalledTimes(1);
                expect(atcoder.contest).toBeCalledWith("aic987");
            });
            test("contest found, with id", async () => {
                atcoder.checkSession = jest.fn(async () => true);
                atcoder.contest = jest.fn(async _ => dummy_contest);
                await contest("aic987", { id: true });
                expect(log).toMatchSnapshot();
                expect(elog).toMatchSnapshot();
                expect(findProjectJSON).toBeCalledTimes(0);
                expect(atcoder.contest).toBeCalledTimes(1);
                expect(atcoder.contest).toBeCalledWith("aic987");
            });
            test("contest not found", async () => {
                atcoder.checkSession = jest.fn(async () => true);
                atcoder.contest = jest.fn(async _ => {
                    throw new Error();
                });
                await contest("aic987654321", {});
                expect(log).toMatchSnapshot();
                expect(elog).toMatchSnapshot();
                expect(findProjectJSON).toBeCalledTimes(0);
                expect(atcoder.contest).toBeCalledTimes(1);
                expect(atcoder.contest).toBeCalledWith("aic987654321");
            });
        });
    });
});
