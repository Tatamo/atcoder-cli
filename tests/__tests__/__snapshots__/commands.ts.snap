// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands contest() contest id is given contest found, with id 1`] = `
"aic987  AtCoder Imaginary Contest 987  https://atcoder.jp/contests/aic987
"
`;

exports[`Commands contest() contest id is given contest found, with id 2`] = `""`;

exports[`Commands contest() contest id is given contest found, without id 1`] = `
"AtCoder Imaginary Contest 987  https://atcoder.jp/contests/aic987
"
`;

exports[`Commands contest() contest id is given contest found, without id 2`] = `""`;

exports[`Commands contest() contest id is given contest not found 1`] = `""`;

exports[`Commands contest() contest id is given contest not found 2`] = `
"contest \\"aic987654321\\" not found.
"
`;

exports[`Commands contest() contest id is not given find project json found, with id 1`] = `
"aic987  AtCoder Imaginary Contest 987  https://atcoder.jp/contests/aic987
"
`;

exports[`Commands contest() contest id is not given find project json found, with id 2`] = `""`;

exports[`Commands contest() contest id is not given find project json found, without id 1`] = `
"AtCoder Imaginary Contest 987  https://atcoder.jp/contests/aic987
"
`;

exports[`Commands contest() contest id is not given find project json found, without id 2`] = `""`;

exports[`Commands contest() contest id is not given project json not found 1`] = `""`;

exports[`Commands contest() contest id is not given project json not found 2`] = `
"contest.acc.json not found.
"
`;

exports[`Commands login() login fail 1`] = `
"login failed
"
`;

exports[`Commands login() login success 1`] = `
"OK
"
`;

exports[`Commands logout() logout 1`] = `
"login session aborted.
"
`;

exports[`Commands session() logged-in 1`] = `
"check login status...
OK
"
`;

exports[`Commands session() logged-in 2`] = `""`;

exports[`Commands session() not logged-in 1`] = `
"check login status...
not login
"
`;

exports[`Commands session() not logged-in 2`] = `""`;
